crates/miden-lib/asm/miden/contracts/auth/multisig.masm

```
@@ -1,24 +1,273 @@
+const.NUMBER_OF_APPROVERS_SLOT=1
+
+# The slot in this component's storage layout where the public keys map is stored.
+const.PUBLIC_KEYS_MAP_SLOT=2
 
-# The slot in this component's storage layout where the public key is stored.
-const.PUBLIC_KEY_SLOT=0
+# The slot in this component's storage layout where executed transactions are stored.
+const.EXECUTED_TXS_SLOT=3
 
-    
-#! Authenticate a transaction using the Falcon signature scheme.
-#! This is a single auth procedure that handles both the addition of a signature and the multisig tx execution.
-#! 
+# Runtime Memory Addresses
+# @dev these can be further optimized, (i do prefer to keep the stack clean and incur the cost of calling mem_{store, load})
+const.TEMP_PK=108
+const.MESSAGE=112
+const.SUCCESSFUL_VERIFICATIONS=116
+const.THRESHOLD=120
+
+# ERRORS
+const.NON_CONSTANT_MSG="NON CONSTANT SIGNATURES"
+const.INVALID_SIGNATURE="SIG DOES NOT MATCH DIGEST"
+
+#! Check if transaction has already been executed and add it to executed transactions for replay protection.
+#!
+#! This procedure ensures that the same transaction cannot be executed multiple times by checking
+#! if the transaction message already exists in the executed transactions map. If it doesn't exist,
+#! it adds the transaction to the map to prevent future replay attacks.
 #!
-#! Inputs:  [TX_HASH]
-#! Outputs: [pad(16)]
+#! Inputs:  [MSG]
+#! Outputs: []
+proc.replay_protection.4
+    # Store the message for later use.
+    # ---------------------------------------------------------------------------------------------
+
+    dupw loc_storew.0 dropw
+    # => [MSG]
+
+    # Check if transaction has already been executed.
+    # ---------------------------------------------------------------------------------------------
+
+    push.EXECUTED_TXS_SLOT
+    # => [index, MSG]
+
+    exec.account::get_map_item
+    # => [LOOKUP_VAL]
+
+    padw
+    # => [EMPTY_WORD, LOOKUP_VAL]
+
+    assert_eqw
```

if this fails, then we've already seen the transaction (someone is trying to replay), correct?
In that case, we should return a meaningful error message